### ISSUE TITLE
`static inline` functions are not supported by header units & modules

### ISSUE THREAD

**Issue by ZacharyHenkel** [dcleblanc/SafeInt#61]
Would it be possible to make `SafeIntExceptionAssert` simply `inline` once again?  

Here's the error as seen in msvc

> safeint3.hpp(430,20): error C2129: static function 'void SafeIntExceptionAssert(void) noexcept' declared but not defined

**Comment by dcleblanc** [comment]
I don't think you have the latest header, which shouldn't matter - this hasn't been changed in years.

I don't understand how 
static inline void SafeIntExceptionAssert() SAFEINT_NOTHROW {}

Could possibly be seen as not defined.

**Comment by ZacharyHenkel** [comment]
`static` functions are local to the TU they are defined in.  Header units are a TU unto themselves.  Thus, the definition of `SafeIntExceptionAssert` does not exist for consumers of the header unit.

**Comment by dcleblanc** [comment]
I believe that's incorrect, they're local to the file they're included in. Having something be static inline means you could have the function defined more than once.

Also, I do not repro this on build, note that the CMakeLists specifies /Wall. Is there some off by default warning that would trigger this?

**Comment by cdacamar** [comment]
The compiler is not required to issue a diagnostic for internal linkage names within a header unit, however using such an internal linkage name is ill-formed as per the standard: [`[basic.link]/2.3`](https://eel.is/c%2B%2Bdraft/basic#link-2.3).  Since the header unit is its own translation unit, the internal linkage function `SafeIntExceptionAssert` is not usable from outside.

The best way to remove this restriction is to make the function `inline`, which would have the same effect as marking it `static`, but also allow the linker to deduplicate instances as well which could decrease binary sizes.

**Comment by dcleblanc** [comment]
Yes, that's fine, but if I'm going to fix something, then I need to be able to reproduce it so that I can add whatever it is to the CMake so it won't regress. I currently get no repro on VS2022, clang, or gcc. I'm not opposed to the fix, I just find it curious that there's not a repro on a fairly recent compiler.

**Comment by cdacamar** [comment]
Here's a minimal repro:

m.h:
```c++
#pragma once
static void f() { }
```

main.cpp:
```c++
import "m.h";

int main() {
  f();
}
```

```
$ cl /std:c++latest /exportHeader m.h
m.h
$ cl /std:c++latest /headerUnit m.h=m.h.ifc main.cpp
main.cpp
D:\msvc\m.h(3): error C2129: static function 'void f(void)' declared but not defined
D:\msvc\m.h(3): note: see declaration of 'f'
```

**Comment by dcleblanc** [comment]
I meant with this header. Also, that function only ever gets called in 4 places inside the header, and never on its own, so your repro isn't accurate. Out of all the compile test code and functionality test code, seems like this error should get hit somewhere.

**Comment by dcleblanc** [comment]
BTW, I thought your use of std:c++latest might be a clue, added a project with that, builds clean.

**Comment by cdacamar** [comment]
Even if the function is called indirectly through a function template or inline function, the static function is reachable:

m.h:
```c++
#pragma once

static void f() { }

inline void g() { f(); }

template <typename>
void h() { f(); }
```

main.cpp:
```c++
import "m.h";

int main() {
  g();
  h<int>();
}
```

We're not directly calling `f`, however it is reached through materializing another definition.  Adding `/std:c++latest` on its own is likely insufficient in your workflows as that will not create a header unit out of the header, which is where the problem manifests.  Note: this will also repro with the, now stable, `/std:c++20` language mode.

**Comment by ZacharyHenkel** [comment]
Sorry we write our own build system so I cannot provide instructions for CMake.  I tried Cameron's minimal repro and it fails with the same error.

Are you testing in two parts to create a header unit via `/exportHeader` and importing it with `/headerUnit m.h=m.h.ifc`?

**Comment by jacobl-at-ms** [comment]
For what its worth, it may not be possible to do this in

**Comment by jacobl-at-ms** [comment]
I tried this set of minimum repro steps based on the info here and couldn't repro the failure either.
```
C:\Users\jacobl\source\repos\SafeInt>type example.cpp
import "SafeInt.hpp";

int main(int argc)
{
    SafeInt<int> x(argc);
    return x;
}
C:\Users\jacobl\source\repos\SafeInt>cl /std:c++latest /nologo /Wall /wd4668 /EHsc /exportHeader SafeInt.hpp
SafeInt.hpp

C:\Users\jacobl\source\repos\SafeInt>cl /std:c++latest /nologo /Wall /wd4668 /EHsc /headerUnit SafeInt.hpp=SafeInt.hpp.ifc example.cpp
```
@ZacharyHenkel  or @cdacamar can you point out what command line option to CL would trigger the issue?   Or is this going to require a specific version of the compiler?  The version of cl I had is:
```
C:\Users\jacobl\source\repos\SafeInt>cl
Microsoft (R) C/C++ Optimizing Compiler Version 19.39.33519 for x86
```

**Comment by dcleblanc** [comment]
That makes sense - you can't throw initializing a SafeInt<int> with an int, nor when unboxing it to return an int. It appears that cmake does support this, but only in the most recent versions, so a non-hacky fix will take some effort to make it part of the normal build-test runs.

Thanks for providing the repro, I'll probably create an interim fix in a branch, and I still need to test that neither gcc or clang are going to do anything unexpected, else I might have to hide the change behind a feature test.
